### ISSUE TITLE
whitebox_tools errors if no \ in the input file path (on Windows)

### ISSUE THREAD

**Issue by AlisterH** [jblindsay/whitebox-geospatial-analysis-tools#44]
Hi,
On Windows it is normally possible to use either back or forward slashes in file paths.
Whitebox_tools produces an error if there is not at least one Windows style \ in the input file path.
This is a problem for the whitebox_tools plugin for QGIS (as per below), but you can reproduce the issue just running from a command prompt.
````
WhiteBox Tools command:
whitebox_tools --run=AbsoluteValue --input="C:/Users/alistair/AppData/Local/Temp/processing_d685dfb1b3cf4ac2b834a7914e97c281/6dba484805634d64a81fedc42787c19a/Terrain.tif" --output="C:/Users/alistair/AppData/Local/Temp/processing_d685dfb1b3cf4ac2b834a7914e97c281/92d44f5f766f41f083829db94e470d1f/output.tif" -v
WhiteBox Tools command output:
**************************** 
* Welcome to AbsoluteValue * 
**************************** 
Reading data... 
thread 'main' panicked at 'called `Result::unwrap()` on an `Err` value: Error { repr: Os { code: 123, message: "The filename, directory name, or volume label syntax is incorrect." } }', src\libcore\result.rs:906:4 
note: Run with `RUST_BACKTRACE=1` for a backtrace. 
Execution completed in 2.75 seconds
Results:
{'output': <QgsProcessingOutputLayerDefinition {'sink':C:/Users/alistair/AppData/Local/Temp/processing_d685dfb1b3cf4ac2b834a7914e97c281/92d44f5f766f41f083829db94e470d1f/output.tif, 'createOptions': {'fileEncoding': 'System'}}>}

Loading resulting layers
The following layers were not correctly generated.
C:/Users/alistair/AppData/Local/Temp/processing_d685dfb1b3cf4ac2b834a7914e97c281/92d44f5f766f41f083829db94e470d1f/output.tif
You can check the 'Log Messages Panel' in QGIS main window to find more information about the execution of the algorithm.
````
original ticket at processing-whitebox: https://github.com/alexbruy/processing-whitebox/issues/4

**Comment by jblindsay** [comment]
I don't use MS Windows regularly. Is this common to use either the backslash or the forward slash in file paths? It was my understanding that only the backslash (\) was used for Windows file paths. How is it that you are specifying a file path with forward slashes to WhiteboxTools? Is it that the QGIS WhiteboxTools plugin is automatically using forward slashes? Are you able to specify the proper file path separator character for the system? I have tested WhiteboxTools using the Python based wb_tools.py GUI on Windows and it works fine, i.e. it specifies the correct file path character.

**Comment by jblindsay** [comment]
Yes, but if we're talking about escape characters in a file path then that's a different story altogether. WhiteboxTools should have no issue with that. What AlisterH is talking about is a Windows file path that only uses forward slashes (/) as the path separator instead of the usual backslash (\\). The issue is that every tool in WhiteboxTools performs a bit of quality control on the input parameters. Because it is common not to specify the fully qualified file name (path included), then each tool will check to see if an input file name contains at least one character that is the system path separator. If not, it appends the working directory:

```
let sep: String = path::MAIN_SEPARATOR.to_string();
if !input_file.contains(&sep) {
     input_file = format!("{}{}", working_directory, input_file);
}
```

If you are on a Windows machine and you specify a fully qualified file path but it uses forward slashes only, it will append the working directory and the path will be specified twice. This is an issue, but given that "path::MAIN_SEPARATOR.to_string();" will always return '\\' on Windows, I'm just not sure how to avoid that. I'm going to have to think about it further.

**Comment by jblindsay** [comment]
I believe that I have fixed this issue with the release of v0.3.1. First, I have fixed the WhiteboxTools Runner Python script such that before calling WhiteboxTools, it normalizes any pathnames. Secondly, within the WhiteboxTools Rust code, I have altered each of the tools that use the following logic:

```Rust
if !input_file.contains(&sep) {
     input_file = format!("{}{}", working_directory, input_file);
}
```

to do this instead:

```Rust
if !input_file.contains(&sep) && !input_file.contains("/") {
     input_file = format!("{}{}", working_directory, input_file);
}
```
That way, even if a user is on Windows and specifies input file names with a forward slash as a path separator, the tool will not append a second working directory string. 

I'm marking this as closed. If you find that the error persists, please re-open.

**Comment by AlisterH** [comment]
For the record, this is fixed in the current Windows download.
